### ISSUE TITLE
Expose Open.stream(source, stream)

### ISSUE THREAD

**Issue by jeremychone** [ZJONSSON/node-unzipper#199]
## Context

We have a case where we need to extract in parallel to upload to some cloud buckets (S3, Google Cloud Storage, and therfore get the `CentralDirectory`. The `Open.s3` is good since it supports stream, but for the other data sources, the `Open.buffer` is not an option as the zip file can be quite big. 

## Proposed Solution

The first option, would be to provide `Open.s3` simiilar function for google and azure, but the list might grow, and one could argue the specitific of the data source should not be the responsibility of an unzipper library. Also, even in the case of `S3` we use our own wrapper, than, now we need to unwrap the s3 client to use `Open.s3`

The preferred option, would be to expose the `unzipper/Open/directory` functions as `Open.stream(source,options): Promise<CentralDirectory>` where it is the caller reponsibility provide `source.size` and `source.stream`. This `directory` interface is very simple and clean, and exposing it will bring great flexibility. 

I am happy to do a pull request if this help.

**Comment by jaapvanblaaderen** [comment]
@jeremychone I've had a similar issue where I want to have more control over the S3 options. I basically implemented your idea and created a PR for it. If you have time, let me know what you think, I guess it's quite a useful addition to this library.

**Comment by jaapvanblaaderen** [comment]
@ZJONSSON Could you have a look at the changes I've pushed in https://github.com/ZJONSSON/node-unzipper/pull/223? Please let me know if you think this adds value to this library.

Thanks!

**Comment by jeremychone** [comment]
@jaapvanblaaderen First, sorry, I saw the issue, wanted to respond and got distracted. It looks ok. 

I assume I could do something like 

```js
const customSource = {
stream: () => { fs.createReadStream(archive) },
size: 12344,
}

or even just 
const stream = fs.createReadStream(archive);
const size = sizeAlreadyComputed; 
const customSource = {stream, size}

Open.custom(customeSource);

```

Right? this way, if I have my stream and size already computed, which I assume is the two minimum things you need to have, I can just pass them along. Adding the promise for size, and function for the stream are nice to have, but this latest simple form would be what I would expect.

**Comment by jaapvanblaaderen** [comment]
Hi @jeremychone, almost, yes. The `size` function always expects to return a Promise returning a number. The `stream` function needs to implement an `offset` and a `length` parameter. The data from these functions are internally used to read chucks of data from the zip file (finding the central directory and extracting entries).

My Open.custom implementation basically passes the customSource to the directory function and exposes that method without any modification of the data.

In the code, you can see two example usages:
* Unit test: [Reading a zip from disk](https://github.com/WoodWing/node-unzipper/blob/open-custom/test/openCustom.js)
* Readme: [Reading a zip from Google Cloud Storage](https://github.com/WoodWing/node-unzipper/tree/open-custom#opencustomsource-options).

Also interesting to see are the current [Open implementation](https://github.com/WoodWing/node-unzipper/blob/open-custom/lib/Open/index.js), that shows how the size and stream methods can be used in various use cases.

**Comment by jeremychone** [comment]
ok, it might work. In our case, we have the size as a number, so, we can wrap it back up as a promise but would be nice to be able to pass a number as well. But no big deal.  

I am not sure I fully understand why stream cannot just be a stream, rather than doing a function with the offset/end, but your readme is clear, so, just a matter of copy/paste. I think in the past, with the CentralDirectory, we just pass the stream and size, and it worked. 

Anyway, this would work, thanks a lot.
